by rahul kumar

github - https://github.com/Rahul5914

youtube - https://youtube.com/@mixupengineer?si=fWNP63I5MDtUH9VZ

linkedin - https://www.linkedin.com/in/rahul-kumar-3829b322a/

instagram -@therahul5914

Importing the dependencies

In [ ]:
import numpy as np
import pandas as pd
import difflib   # i/p give movie and this lib compare to closest match from data, (most closest match name given by user)
from sklearn.feature_extraction.text import TfidfVectorizer# tfidfviectr, textual data to numerical value(feature vector)like in spam mail
from sklearn.metrics.pairwise import cosine_similarity# for highset imila score for similar movivies ,its not ml ago

Data Collection and Pre-Processing

In [ ]:
# loading the data from the csv file to apandas dataframe
movies_data = pd.read_csv('/content/movies.csv', on_bad_lines='skip')
#or
#movies_data = pd.read_csv('/content/movies.csv', error_bad_lines=False)

This inconsistency often arises due to:

**Extra delimiters (commas) within data fields:** If a data field contains commas without being properly quoted, Pandas may interpret them as column separators, leading to more columns than expected in that specific row.

**Missing values**: Rows with fewer columns than the initially inferred schema (24 in this case) can also trigger this error if not properly handled.


# loading the data from the csv file to apandas dataframe
movies_data = pd.read_csv('/content/movies.csv', on_bad_lines='skip')
#or
#movies_data = pd.read_csv('/content/movies.csv', error_bad_lines=False)

In [ ]:
# loading the data from the csv file to apandas dataframe
#movies_data = pd.read_csv('/content/movies.csv') give parser error solver by upper code




In [ ]:
# printing the first 5 rows of the dataframe
movies_data.head()

index     budget  ...                                               crew           director
0      0  237000000  ...  [{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...      James Cameron
1      1  300000000  ...  [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...     Gore Verbinski
2      2  245000000  ...  [{'name': 'Thomas Newman', 'gender': 2, 'depar...         Sam Mendes
3      3  250000000  ...  [{'name': 'Hans Zimmer', 'gender': 2, 'departm...  Christopher Nolan
4      4  260000000  ...  [{'name': 'Andrew Stanton', 'gender': 2, 'depa...     Andrew Stanton

[5 rows x 24 columns]

In [ ]:
# number of rows and columns in the data frame

movies_data.shape

(2936, 24)

In [ ]:
# selecting the relevant features for recommendation

selected_features = ['genres','keywords','tagline','cast','director']
print(selected_features)

['genres', 'keywords', 'tagline', 'cast', 'director']


In [ ]:
# replacing the null valuess with null string

for feature in selected_features:
  movies_data[feature] = movies_data[feature].fillna('')

In [ ]:
# combining all the 5 selected features

combined_features = movies_data['genres']+' '+movies_data['keywords']+' '+movies_data['tagline']+' '+movies_data['cast']+' '+movies_data['director']

In [ ]:
print(combined_features)

0       Action Adventure Fantasy Science Fiction cultu...
1       Adventure Fantasy Action ocean drug abuse exot...
2       Action Adventure Crime spy based on novel secr...
3       Action Crime Drama Thriller dc comics crime fi...
4       Action Adventure Science Fiction based on nove...
                              ...                        
2931    Action Crime Thriller united states\u2013mexic...
2932    Comedy Romance  A newlywed couple's honeymoon ...
2933    Comedy Drama Romance TV Movie date love at fir...
2934      A New Yorker in Shanghai Daniel Henney Eliza...
2935    Documentary obsession camcorder crush dream gi...
Length: 2936, dtype: object


In [ ]:
# converting the text data to feature vectors

vectorizer = TfidfVectorizer()

In [ ]:
feature_vectors = vectorizer.fit_transform(combined_features)

In [ ]:
print(feature_vectors)

  (0, 150)	0.08514895551490012
  (0, 211)	0.09569524321624796
  (0, 4405)	0.11870032983164201
  (0, 11428)	0.10528898106487948
  (0, 4541)	0.10518269307924367
  (0, 3064)	0.19974661168400182
  (0, 2557)	0.20642265047507175
  (0, 4868)	0.16376098138367745
  (0, 12092)	0.3491585315962019
  (0, 13914)	0.12757510908439554
  (0, 2702)	0.24431886772303396
  (0, 12000)	0.2148582089677792
  (0, 4141)	0.23714433698729528
  (0, 12819)	0.07297753708760794
  (0, 14250)	0.13339457015894768
  (0, 9427)	0.09189949654994621
  (0, 9688)	0.26660499574255
  (0, 11217)	0.153225147567914
  (0, 14259)	0.23714433698729528
  (0, 14485)	0.19779740360694614
  (0, 11191)	0.21824576936765805
  (0, 11828)	0.2117937878343536
  (0, 13977)	0.2040399245533557
  (0, 12272)	0.15275576570796295
  (0, 7364)	0.2220327397035179
  :	:
  (2934, 312)	0.18427520182952378
  (2934, 14466)	0.28301718577995194
  (2934, 4045)	0.25261093537312324
  (2934, 11633)	0.27269338639194585
  (2934, 11068)	0.28301718577995194
  (2934, 14368)	

Cosine Similarity

In [ ]:
# getting the similarity scores using cosine similarity

similarity = cosine_similarity(feature_vectors)

In [ ]:
print(similarity)

[[1.         0.07891604 0.04053147 ... 0.         0.         0.        ]
 [0.07891604 1.         0.03651211 ... 0.03555329 0.         0.        ]
 [0.04053147 0.03651211 1.         ... 0.         0.05358917 0.        ]
 ...
 [0.         0.03555329 0.         ... 1.         0.         0.02730171]
 [0.         0.         0.05358917 ... 0.         1.         0.        ]
 [0.         0.         0.         ... 0.02730171 0.         1.        ]]


In [ ]:
print(similarity.shape)

(2936, 2936)


Getting the movie name from the user

In [ ]:
# getting the movie name from the user

movie_name = input(' Enter your favourite movie name : ')

 Enter your favourite movie name : iron man


In [ ]:
# creating a list with all the movie names given in the dataset
list_of_all_titles = movies_data['title'].astype(str).tolist()
print(list_of_all_titles)

# finding the close match for the movie name given by the user
find_close_match = difflib.get_close_matches(movie_name, list_of_all_titles)
print(find_close_match)

['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre', 'The Dark Knight Rises', 'John Carter', 'Spider-Man 3', 'Tangled', 'Avengers: Age of Ultron', 'Harry Potter and the Half-Blood Prince', 'Batman v Superman: Dawn of Justice', 'Superman Returns', 'Quantum of Solace', "Pirates of the Caribbean: Dead Man's Chest", 'The Lone Ranger', 'Man of Steel', 'The Chronicles of Narnia: Prince Caspian', 'The Avengers', 'Pirates of the Caribbean: On Stranger Tides', 'Men in Black 3', 'The Hobbit: The Battle of the Five Armies', 'The Amazing Spider-Man', 'Robin Hood', 'The Hobbit: The Desolation of Smaug', 'The Golden Compass', 'King Kong', 'Titanic', 'Captain America: Civil War', 'Battleship', 'Jurassic World', 'Skyfall', 'Spider-Man 2', 'Iron Man 3', 'Alice in Wonderland', 'X-Men: The Last Stand', 'Monsters University', 'Transformers: Revenge of the Fallen', 'Transformers: Age of Extinction', 'Oz: The Great and Powerful', 'The Amazing Spider-Man 2', 'TRON: Legacy', 'Cars 2', 'Green Lant

In [ ]:
# creating a list with all the movie names given in the dataset

#list_of_all_titles = movies_data['title'].tolist()
#print(list_of_all_titles)

['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre', 'The Dark Knight Rises', 'John Carter', 'Spider-Man 3', 'Tangled', 'Avengers: Age of Ultron', 'Harry Potter and the Half-Blood Prince', 'Batman v Superman: Dawn of Justice', 'Superman Returns', 'Quantum of Solace', "Pirates of the Caribbean: Dead Man's Chest", 'The Lone Ranger', 'Man of Steel', 'The Chronicles of Narnia: Prince Caspian', 'The Avengers', 'Pirates of the Caribbean: On Stranger Tides', 'Men in Black 3', 'The Hobbit: The Battle of the Five Armies', 'The Amazing Spider-Man', 'Robin Hood', 'The Hobbit: The Desolation of Smaug', 'The Golden Compass', 'King Kong', 'Titanic', 'Captain America: Civil War', 'Battleship', 'Jurassic World', 'Skyfall', 'Spider-Man 2', 'Iron Man 3', 'Alice in Wonderland', 'X-Men: The Last Stand', 'Monsters University', 'Transformers: Revenge of the Fallen', 'Transformers: Age of Extinction', 'Oz: The Great and Powerful', 'The Amazing Spider-Man 2', 'TRON: Legacy', 'Cars 2', 'Green Lant

The error "TypeError: object of type 'float' has no len()" is occurring within the difflib.get_close_matches function. This function expects the input sequences (in this case, the movie titles in list_of_all_titles) to be strings or sequences of hashable objects. The traceback suggests that at least one element within list_of_all_titles is a float, which is not a valid input type for this function. This could be due to missing or improperly formatted data in the 'title' column of your movies_data DataFrame.

In [ ]:
# finding the close match for the movie name given by the user

#find_close_match = difflib.get_close_matches(movie_name, list_of_all_titles)
#print(find_close_match)

# error -- 'float' has no len ,solved  by above code

In [ ]:
close_match = find_close_match[0]
print(close_match)

Iron Man


In [ ]:
# finding the index of the movie with title

index_of_the_movie = movies_data[movies_data.title == close_match]['index'].values[0]
print(index_of_the_movie)

68


In [ ]:
# getting a list of similar movies

similarity_score = list(enumerate(similarity[index_of_the_movie]))
print(similarity_score)

[(0, 0.036808113027465186), (1, 0.057759057692149), (2, 0.015917183629317242), (3, 0.007690462709730372), (4, 0.03615358361002035), (5, 0.01627984643143107), (6, 0.08056419429631693), (7, 0.23903287997934555), (8, 0.009250228718698563), (9, 0.07719370999385816), (10, 0.07880908388154942), (11, 0.013919234295953715), (12, 0.016041625287786156), (13, 0.014751707694882203), (14, 0.09761333816520024), (15, 0.008534090685360526), (16, 0.2281852480643192), (17, 0.01529469699603088), (18, 0.04484698144212966), (19, 0.08091946475070273), (20, 0.07672677187298071), (21, 0.013224288416633067), (22, 0.007984524444896486), (23, 0.007681330971113011), (24, 0.014909426394609702), (25, 0.0), (26, 0.21329439069718145), (27, 0.034101235979107136), (28, 0.06486889747367303), (29, 0.0162913920712559), (30, 0.0783847731509894), (31, 0.3105316888558051), (32, 0.030130891415050598), (33, 0.12829035122936225), (34, 0.0), (35, 0.039516178764570524), (36, 0.03536839990793246), (37, 0.009135698997401872), (38, 

In [ ]:
len(similarity_score)

2936

In [ ]:
# sorting the movies based on their similarity score

sorted_similar_movies = sorted(similarity_score, key = lambda x:x[1], reverse = True)
print(sorted_similar_movies)

[(68, 1.0000000000000002), (79, 0.4050702719841016), (31, 0.3105316888558051), (7, 0.23903287997934555), (16, 0.2281852480643192), (26, 0.21329439069718145), (85, 0.20228878829293195), (416, 0.19295888850762086), (1756, 0.1660960245203289), (2534, 0.15270311990170973), (64, 0.14904667526073312), (680, 0.14583434720874816), (408, 0.14466764226684184), (46, 0.14000034404866396), (335, 0.13941161961265305), (403, 0.13737137125264642), (365, 0.13537630286850355), (94, 0.1333208718894716), (360, 0.13072625702390184), (33, 0.12829035122936225), (526, 0.12509236118355796), (481, 0.11933117485981667), (372, 0.11271494223885653), (784, 0.11239409580603917), (38, 0.11045371794512863), (719, 0.10697228551990134), (356, 0.10640949575136496), (1576, 0.10292632606523858), (1518, 0.10140087612672725), (1013, 0.10023057824839895), (173, 0.09961157309121477), (328, 0.09841397313059104), (1008, 0.09829293858066167), (1249, 0.09809754627332906), (14, 0.09761333816520024), (1299, 0.09757915441614), (1519,

In [ ]:
# print the name of similar movies based on the index

print('Movies suggested for you : \n')

i = 1

for movie in sorted_similar_movies:
  index = movie[0]
  title_from_index = movies_data[movies_data.index==index]['title'].values[0]
  if (i<30):
    print(i, '.',title_from_index)
    i+=1

Movies suggested for you : 

1 . Iron Man
2 . Iron Man 2
3 . Iron Man 3
4 . Avengers: Age of Ultron
5 . The Avengers
6 . Captain America: Civil War
7 . Captain America: The Winter Soldier
8 . Ant-Man
9 . Made
10 . The Helix... Loaded
11 . X-Men: Apocalypse
12 . X2
13 . The Incredible Hulk
14 . X-Men: Days of Future Past
15 . X-Men: First Class
16 . Captain America: The First Avenger
17 . G-Force
18 . Guardians of the Galaxy
19 . Thor: The Dark World
20 . X-Men: The Last Stand
21 . Duets
22 . Southland Tales
23 . The Last Airbender
24 . The Good Night
25 . The Amazing Spider-Man 2
26 . Fantastic Four
27 . X-Men Origins: Wolverine
28 . Mallrats
29 . March or Die


Movie Recommendation Sytem

In [ ]:
movie_name = input(' Enter your favourite movie name : ')

list_of_all_titles = movies_data['title'].astype(str).tolist()

find_close_match = difflib.get_close_matches(movie_name, list_of_all_titles)

close_match = find_close_match[0]

# Get the index from the movies_data DataFrame using the 'title' column
index_of_the_movie = movies_data[movies_data.title == close_match].index[0]

# Ensure the index is within the bounds of the similarity matrix
index_of_the_movie = min(index_of_the_movie, similarity.shape[0] - 1)

similarity_score = list(enumerate(similarity[index_of_the_movie]))

sorted_similar_movies = sorted(similarity_score, key = lambda x:x[1], reverse = True)

print('Movies suggested for you : \n')

i = 1

for movie in sorted_similar_movies:
  index = movie[0]
  # Ensure the index is within the bounds of the DataFrame
  index = min(index, movies_data.shape[0] - 1)
  title_from_index = movies_data[movies_data.index==index]['title'].values[0]
  if (i<30):
    print(i, '.',title_from_index)
    i+=1

 Enter your favourite movie name : Avatar
Movies suggested for you : 

1 . Avatar
2 . Alien
3 . Aliens
4 . Guardians of the Galaxy
5 . Star Trek Beyond
6 . Star Trek Into Darkness
7 . Cargo
8 . Gravity
9 . Jason X
10 . Trekkies
11 . The Helix... Loaded
12 . Lockout
13 . Machete Kills
14 . Star Wars: Clone Wars: Volume 1
15 . Clash of the Titans
16 . Star Trek
17 . Planet of the Apes
18 . Terminator Salvation
19 . The Terminator
20 . Silent Running
21 . Imaginary Heroes
22 . In the Shadow of the Moon
23 . Terminator Genisys
24 . Men in Black II
25 . John Carter
26 . Jupiter Ascending
27 . Crocodile Dundee
28 . The Ice Pirates
29 . Cedar Rapids


The error IndexError: index 4267 is out of bounds for axis 0 with size 2936 arises because the index_of_the_movie variable holds a value (4267) that exceeds the maximum index of the similarity array.

The similarity array, representing the cosine similarity between movies, likely has dimensions of 2936 x 2936, indicating it holds similarity scores for 2936 movies. When trying to access similarity[index_of_the_movie] with index_of_the_movie being 4267, it attempts to access an element beyond the bounds of the array, leading to the IndexError.

This inconsistency between the index_of_the_movie value and the size of the similarity array might be due to the way index_of_the_movie is derived. It's likely obtained from the 'index' column of the movies_data DataFrame. This column might contain values that do not correspond directly to the indices of the similarity array, potentially due to data preprocessing or filtering steps

In [ ]:
# movie_name = input(' Enter your favourite movie name : ')

# list_of_all_titles = movies_data['title'].tolist()

# find_close_match = difflib.get_close_matches(movie_name, list_of_all_titles)

# close_match = find_close_match[0]

# index_of_the_movie = movies_data[movies_data.title == close_match]['index'].values[0]

# similarity_score = list(enumerate(similarity[index_of_the_movie]))

# sorted_similar_movies = sorted(similarity_score, key = lambda x:x[1], reverse = True)

# print('Movies suggested for you : \n')

# i = 1

# for movie in sorted_similar_movies:
#   index = movie[0]
#   title_from_index = movies_data[movies_data.index==index]['title'].values[0]
#   if (i<30):
#     print(i, '.',title_from_index)
#     i+=1

#this cell code giving index error ollver by above code